### PR TITLE
Added upper limit of the instance count in prediction score labeling suggestion method

### PR DESCRIPTION
**Pull request by hsingchien**
### Description
PR 2/3 for https://github.com/talmolab/sleap/discussions/977
Renamed original instance num input to `instance_limit_lower`. Added `instance_limit_upper` to the method. Updated suggestion config file accordingly. User can now have finer control over the number of instances in generating prediction score based labeling suggestion. 

---

I found it useful to quickly generate frame list with only certain number of visible predicted instances. I am working on a 2-animal project, having 1 instance means failure on the other animal. 
This can be achieved by a small tweak on <prediction_score> suggestion method. I renamed original input <num_instance> to <instance_at_least> added another input <instance_no_more_than>. Now I can just set <predition_score> to 100 and both instance upperbound and lowerbound to be 1, and label suggestion will give back all frames with precisely 1 predicted instance. This obviously enables other frame filtering possibilities as well.

_Originally posted by @hsingchien in https://github.com/talmolab/sleap/discussions/977_

### Types of changes

- [ ] Bugfix
- [x] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
[list open issues here]

### Outside contributors checklist

- [x] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [x] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -18,6 +18,7 @@ dependencies:
   # - sleap::pyside2=5.14.1
   - qtpy>=2.0.1
   - conda-forge::pip!=22.0.4
+  - pip
```
Comment (roomrys) [inline environment.yml:21]: Did you experience any issues with this removed?

**Review by roomrys — CHANGES_REQUESTED** [review]
A small correction that for an edge case and then we can also add some tests in the tests/gui/test_suggestions.py file. I wrote a sample test that we can add to:
```py
def test_limits_prediction_score(centered_pair_predictions: Labels):
    """Testing suggestion generation using instance limits and prediction score."""
    labels = centered_pair_predictions
    score_limit = 20
    instance_lower_limit = 3
    instance_upper_limit = 3

    # Generate suggestions
    suggestions = VideoFrameSuggestions.suggest(
        labels=labels,
        params={
            "videos": labels.videos,
            "method": "prediction_score",
            "score_limit": score_limit,
            "instance_limit_upper": instance_upper_limit,
            "instance_limit_lower": instance_lower_limit,
        },
    )

    # Confirming every suggested frame meets criteria
    for sugg in suggestions:
        lf = labels.get((sugg.video, sugg.frame_idx))
        pred_instances = [
            inst for inst in lf.instances_to_show if isinstance(inst, PredictedInstance)
        ]
        n_instance_below_score = np.nansum(
            [True for inst in pred_instances if inst.score < score_limit]
        )
        assert n_instance_below_score >= instance_lower_limit
        assert n_instance_below_score <= instance_upper_limit
```
The limits are set score limit is set s.t. we only get 6 suggestions returned (for speed of the assertion loop at the end).

File: sleap/gui/suggestions.py [hunk 2]
```diff
@@ -195,29 +203,35 @@ def prediction_score(
...
+
+            if (
+                n_qualified_instance >= instance_limit_lower
+                and n_qualified_instance <= instance_limit_upper
+            ):
+                # idxs saves qualified frame index at corresponding entry, otherwise the entry is 0
+                idxs[i] = lf.frame_idx
+
+        # Finds non-zero entries in idxs
+        result = sorted(idxs[idxs > 0].tolist())
```
[the remark this answers is not in the document]
Comment (hsingchien) [reply sleap/gui/suggestions.py]: Changed accordingly! Thank you!

File: sleap/gui/suggestions.py [hunk 3]
```diff
@@ -195,29 +203,35 @@ def prediction_score(
...
+        labels: "Labels",
+        score_limit: float,
+        instance_limit_upper: int,
+        instance_limit_lower: int,
     ):
         lfs = labels.find(video)
         frames = len(lfs)
-        idxs = np.ndarray((frames), dtype="int")
-        scores = np.full((frames, instance_limit), 100.0, dtype="float")
+        idxs = np.zeros((frames), dtype="int")
```
[the remark this answers is not in the document]
Comment (hsingchien) [reply sleap/gui/suggestions.py]: Corrected!

File: environment.yml (hunk 1 above)
Comment (hsingchien) [reply environment.yml:21]: Old yaml gave infinite warnings suggesting some packages require `pip`, without `pip` installation cannot be done. Adding pip solved the problem, so was environment_build.yml.  environment_m1.yml has pip line though.

**Comment by hsingchien** [comment]
I have corrected the comments in suggestions.py . Also added the sample test from @roomrys to confirm all the suggested frames meet the criteria (corrected  `inst.score < score_limit` in the original script to `inst.score <= score_limit`). Also added another test to confirm all the qualified frames in target videos are successfully captured.

File: tests/gui/test_suggestions.py [hunk 4]
```diff
@@ -366,3 +369,62 @@ def test_append_suggestions(small_robot_3_frame_vid: Video, stickman: Skeleton):
...
+        n_instance_below_score = np.nansum(
+            [True for inst in pred_instances if inst.score <= score_limit]
+        )
+        assert n_instance_below_score >= instance_lower_limit
+        assert n_instance_below_score <= instance_upper_limit
+
+    # Confirming all frames meeting the criteria are captured
+    def check_all_predicted_instances(sugg, labels):
+        for video in labels.videos:
+            lfs = labels.find(video)
```
Comment (roomrys) [inline tests/gui/test_suggestions.py:408]: We can remove this outer for loop as there is only a single video in the `centered_pair_predictions` project.
```suggestion
lfs = labels.labeled_frames
```

**Review by roomrys — COMMENTED** [review]
Minor suggestion to speed-up the tests, then I believe we are good to go.

File: tests/gui/test_suggestions.py (hunk 4 above)
Comment (hsingchien) [reply tests/gui/test_suggestions.py]: Done! Also corrected another mistake I made earlier

File: sleap/gui/suggestions.py [hunk 5]
```diff
@@ -195,29 +203,36 @@ def prediction_score(
...
+            # Scores from visible instances in frame
+            pred_fs = lf.instances_to_show
+            frame_scores = np.array(
+                [inst.score for inst in pred_fs if hasattr(inst, "score")]
+            )
+            # Gets the number of instances with scores lower than <score_limit>
+            n_qualified_instance = np.nansum(frame_scores <= score_limit)
+
+            if (
+                n_qualified_instance >= instance_limit_lower
```
Comment (talmo) [inline sleap/gui/suggestions.py:228]: `instance_limit_lower` has a default of 0, which will evaluate to `True` for empty frames I think. Like:

```py
frame_scores = np.array([])  # == no predicted instances in frame
n_qualified_instance = np.nansum(frame_scores <= score_limit)  # == 0
```

Do we maybe want the lower limit to be 1 or switch to using `>` instead of `>=`?

Comment (hsingchien) [reply sleap/gui/suggestions.py:228]: That's a good point. I guess one of the uses of this feature is to quickly locate the frames where prediction failed, and that would include empty frames (e.g. after a whole video prediction). Therefore, I prefer to keep the option open to users

**Review by roomrys — APPROVED** [review]

**Comment by hsingchien** [comment]
Regarding to @talmo 's point earlier, I realized that setting lower limit to 0 will also give out frames with user instances and no predicted instances, which is likely not desirable if user only wants to get frames failed in prediction. A potential solution would be to give user an option to narrow the searching scope to unlabeled frames. This could be a dropdown menu like `target video` that can be applied to all methods or a button/menu command to remove labeled frames from labeling suggestions.
